File: pkg/integrations/github/components/checks/on_check_run.go
package checks

import (
	"encoding/json"
	"fmt"
	"net/http"
	"slices"
	"strconv"

	"github.com/mitchellh/mapstructure"
	"github.com/superplanehq/superplane/pkg/configuration"
	"github.com/superplanehq/superplane/pkg/core"
	"github.com/superplanehq/superplane/pkg/integrations/github/common"
)

type OnCheckRun struct{}

type OnCheckRunConfiguration struct {
	Repository       string                    `json:"repository" mapstructure:"repository"`
	Statuses         []string                  `json:"statuses" mapstructure:"statuses"`
	Conclusions      []string                  `json:"conclusions" mapstructure:"conclusions"`
	Names            []configuration.Predicate `json:"names" mapstructure:"names"`
	Branches         []configuration.Predicate `json:"branches" mapstructure:"branches"`
	PullRequestsOnly bool                      `json:"pullRequestsOnly" mapstructure:"pullRequestsOnly"`
}

func (t *OnCheckRun) Name() string {
	return "github.onCheckRun"
}

func (t *OnCheckRun) Label() string {
	return "On Check Run"
}

func (t *OnCheckRun) Description() string {
	return "Listen to GitHub Checks API check run events"
}

func (t *OnCheckRun) Documentation() string {
	return `The On Check Run trigger starts a workflow execution when a GitHub Checks API check run changes.

GitHub check runs are created by GitHub Apps and power many pull request checks such as Cloudflare Pages, DCO, Sourcery, and GitHub Actions check runs. They are separate from legacy Commit Statuses API statuses. Use On Commit Status for legacy commit status events.

## Use Cases

- **Pull request automation**: React when Checks API checks complete or fail
- **Quality gates**: Continue workflows only after a specific check run reaches the expected state
- **Notifications**: Notify teams when app-provided checks fail or require action
- **Merge orchestration**: Combine this trigger with List Check Runs For Ref to decide whether all checks are green

## Configuration

- **Repository**: Select the GitHub repository to monitor
- **Statuses**: Select check run statuses to listen for (queued, in_progress, completed, requested, waiting, pending)
- **Conclusions** *(optional)*: Filter completed check runs by conclusion
- **Names** *(optional)*: Filter check run names using predicates, e.g. equals "DCO" or matches "Cloudflare.*"
- **Branches** *(optional)*: Filter check suite head branches using predicates
- **Pull requests only**: Emit only check runs that GitHub associates with at least one pull request

## Event Data

Each event includes:
- **action**: The check_run webhook action
- **check_run**: Check run details, including name, status, conclusion, head SHA, app, pull requests, and URLs
- **repository**: Repository information
- **sender**: User or app actor associated with the webhook event

## Webhook Setup

This trigger automatically sets up a GitHub webhook when configured. The webhook is managed by SuperPlane and will be cleaned up when the trigger is removed.

## Notes

GitHub may return an empty pull_requests array for check runs created from forked repository pushes. When Pull requests only is enabled, those fork-originated check runs may be filtered out because GitHub does not include PR metadata in the webhook payload.`
}

func (t *OnCheckRun) Icon() string {
	return "github"
}

func (t *OnCheckRun) Color() string {
	return "gray"
}

func (t *OnCheckRun) Configuration() []configuration.Field {
	return []configuration.Field{
		{
			Name:     "repository",
			Label:    "Repository",
			Type:     configuration.FieldTypeIntegrationResource,
			Required: true,
			TypeOptions: &configuration.TypeOptions{
				Resource: &configuration.ResourceTypeOptions{
					Type:           "repository",
					UseNameAsValue: true,
				},
			},
		},
		{
			Name:     "statuses",
			Label:    "Statuses",
			Type:     configuration.FieldTypeMultiSelect,
			Required: true,
			Default:  []string{"queued", "in_progress", "completed", "requested", "waiting", "pending"},
			TypeOptions: &configuration.TypeOptions{
				MultiSelect: &configuration.MultiSelectTypeOptions{
					Options: []configuration.FieldOption{
						{Label: "Queued", Value: "queued"},
						{Label: "In Progress", Value: "in_progress"},
						{Label: "Completed", Value: "completed"},
						{Label: "Requested", Value: "requested"},
						{Label: "Waiting", Value: "waiting"},
						{Label: "Pending", Value: "pending"},
					},
				},
			},
		},
		{
			Name:        "conclusions",
			Label:       "Conclusions",
			Description: "Optional. Filter completed check runs by conclusion.",
			Type:        configuration.FieldTypeMultiSelect,
			Required:    false,
			Togglable:   true,
			TypeOptions: &configuration.TypeOptions{
				MultiSelect: &configuration.MultiSelectTypeOptions{
					Options: checkRunConclusionOptions(),
				},
			},
		},
		{
			Name:        "names",
			Label:       "Names",
			Description: "Optional. Filter check run names.",
			Type:        configuration.FieldTypeAnyPredicateList,
			Required:    false,
			Togglable:   true,
			TypeOptions: &configuration.TypeOptions{
				AnyPredicateList: &configuration.AnyPredicateListTypeOptions{
					Operators: configuration.AllPredicateOperators,
				},
			},
		},
		{
			Name:        "branches",
			Label:       "Branches",
			Description: "Optional. Filter check suite head branch names.",
			Type:        configuration.FieldTypeAnyPredicateList,
			Required:    false,
			Togglable:   true,
			TypeOptions: &configuration.TypeOptions{
				AnyPredicateList: &configuration.AnyPredicateListTypeOptions{
					Operators: configuration.AllPredicateOperators,
				},
			},
		},
		{
			Name:        "pullRequestsOnly",
			Label:       "Pull requests only",
			Description: "Emit only check runs associated with at least one pull request.",
			Type:        configuration.FieldTypeBool,
			Required:    false,
			Default:     false,
		},
	}
}

func (t *OnCheckRun) Setup(ctx core.TriggerContext) error {
	if err := common.EnsureRepoInMetadata(ctx.Metadata, ctx.Integration, ctx.HTTP, ctx.Configuration); err != nil {
		return err
	}

	var config OnCheckRunConfiguration
	if err := mapstructure.Decode(ctx.Configuration, &config); err != nil {
		return fmt.Errorf("failed to decode configuration: %w", err)
	}

	return ctx.Integration.RequestWebhook(common.WebhookConfiguration{
		EventType:  "check_run",
		Repository: config.Repository,
	})
}

func (t *OnCheckRun) Hooks() []core.Hook {
	return []core.Hook{}
}

func (t *OnCheckRun) HandleHook(ctx core.TriggerHookContext) (map[string]any, error) {
	return nil, nil
}

func (t *OnCheckRun) HandleWebhook(ctx core.WebhookRequestContext) (int, *core.WebhookResponseBody, error) {
	ctx = common.WithWebhookLogger(ctx, t.Name())
	ctx.Logger.Infof("Received GitHub webhook")

	var config OnCheckRunConfiguration
	if err := mapstructure.Decode(ctx.Configuration, &config); err != nil {
		ctx.Logger.Errorf("Failed to decode configuration: %v", err)
		return http.StatusInternalServerError, nil, fmt.Errorf("failed to decode configuration: %w", err)
	}

	eventType := ctx.Headers.Get("X-GitHub-Event")
	if eventType == "" {
		ctx.Logger.Errorf("Missing X-GitHub-Event header")
		return http.StatusBadRequest, nil, fmt.Errorf("missing X-GitHub-Event header")
	}

	if eventType != "check_run" {
		ctx.Logger.Infof("Ignoring event - event type %q is not a check_run event", eventType)
		return http.StatusOK, nil, nil
	}

	code, err := common.VerifySignature(ctx)
	if err != nil {
		ctx.Logger.Errorf("Failed to verify signature: %v", err)
		return code, nil, err
	}

	payload := map[string]any{}
	if err := json.Unmarshal(ctx.Body, &payload); err != nil {
		ctx.Logger.Errorf("Failed to parse request body: %v", err)
		return http.StatusBadRequest, nil, fmt.Errorf("error parsing request body: %v", err)
	}

	if !matchesCheckRunFilters(payload, config) {
		ctx.Logger.Info("Ignoring event - check run did not match configured filters")
		return http.StatusOK, nil, nil
	}

	if err := ctx.Events.Emit("github.checkRun", payload); err != nil {
		ctx.Logger.Errorf("Failed to emit event: %v", err)
		return http.StatusInternalServerError, nil, fmt.Errorf("error emitting event: %v", err)
	}

	return http.StatusOK, nil, nil
}

func (t *OnCheckRun) Cleanup(ctx core.TriggerContext) error {
	return nil
}

func matchesCheckRunFilters(payload map[string]any, config OnCheckRunConfiguration) bool {
	checkRun, ok := payload["check_run"].(map[string]any)
	if !ok {
		return false
	}

	status := extractStringValue(checkRun, "status")
	if !matchesConfiguredCheckRunStatus(status, config.Statuses) {
		return false
	}

	if !matchesConfiguredCheckRunConclusion(status, extractStringValue(checkRun, "conclusion"), config.Conclusions) {
		return false
	}

	if !matchesCheckRunName(extractStringValue(checkRun, "name"), config.Names) {
		return false
	}

	if !matchesCheckRunBranch(checkRun, config.Branches) {
		return false
	}

	if config.PullRequestsOnly && len(extractCheckRunPullRequestNumbers(checkRun)) == 0 {
		return false
	}

	return true
}

func matchesConfiguredCheckRunStatus(status string, allowedStatuses []string) bool {
	if len(allowedStatuses) == 0 {
		return true
	}

	return slices.Contains(allowedStatuses, status)
}

func matchesConfiguredCheckRunConclusion(status string, conclusion string, allowedConclusions []string) bool {
	if len(allowedConclusions) == 0 {
		return true
	}

	if status != "completed" {
		return true
	}

	if conclusion == "" {
		return false
	}

	return slices.Contains(allowedConclusions, conclusion)
}

func matchesCheckRunName(name string, allowedNames []configuration.Predicate) bool {
	if len(allowedNames) == 0 {
		return true
	}

	return configuration.MatchesAnyPredicate(allowedNames, name)
}

func matchesCheckRunBranch(checkRun map[string]any, allowedBranches []configuration.Predicate) bool {
	if len(allowedBranches) == 0 {
		return true
	}

	branch := extractCheckRunBranch(checkRun)
	if branch == "" {
		return false
	}

	return configuration.MatchesAnyPredicate(allowedBranches, branch)
}

func extractCheckRunBranch(checkRun map[string]any) string {
	checkSuite, ok := checkRun["check_suite"].(map[string]any)
	if ok {
		branch := extractStringValue(checkSuite, "head_branch")
		if branch != "" {
			return branch
		}
	}

	return extractFirstPullRequestHeadRef(checkRun["pull_requests"])
}

func extractFirstPullRequestHeadRef(value any) string {
	pullRequests, ok := value.([]any)
	if !ok || len(pullRequests) == 0 {
		return ""
	}

	pullRequest, ok := pullRequests[0].(map[string]any)
	if !ok {
		return ""
	}

	head, ok := pullRequest["head"].(map[string]any)
	if !ok {
		return ""
	}

	return extractStringValue(head, "ref")
}

func extractCheckRunPullRequestNumbers(checkRun map[string]any) []string {
	numbers := extractPullRequestNumbers(checkRun["pull_requests"])
	if len(numbers) > 0 {
		return numbers
	}

	checkSuite, ok := checkRun["check_suite"].(map[string]any)
	if !ok {
		return nil
	}

	return extractPullRequestNumbers(checkSuite["pull_requests"])
}

func extractPullRequestNumbers(value any) []string {
	pullRequests, ok := value.([]any)
	if !ok {
		return nil
	}

	numbers := []string{}
	for _, pullRequestRaw := range pullRequests {
		pullRequest, ok := pullRequestRaw.(map[string]any)
		if !ok {
			continue
		}

		switch number := pullRequest["number"].(type) {
		case float64:
			numbers = append(numbers, strconv.Itoa(int(number)))
		case string:
			numbers = append(numbers, number)
		}
	}

	return numbers
}

func extractStringValue(data map[string]any, key string) string {
	value, ok := data[key]
	if !ok {
		return ""
	}

	text, ok := value.(string)
	if !ok {
		return ""
	}

	return text
}

func checkRunConclusionOptions() []configuration.FieldOption {
	return []configuration.FieldOption{
		{Label: "Success", Value: "success"},
		{Label: "Failure", Value: "failure"},
		{Label: "Neutral", Value: "neutral"},
		{Label: "Cancelled", Value: "cancelled"},
		{Label: "Skipped", Value: "skipped"},
		{Label: "Timed Out", Value: "timed_out"},
		{Label: "Action Required", Value: "action_required"},
		{Label: "Stale", Value: "stale"},
	}
}

File: web_src/src/pages/app/mappers/github/on_check_run.ts
import { getColorClass, getBackgroundColorClass } from "@/lib/colors";
import type { MetadataItem } from "@/ui/metadataList";
import type { TriggerEventContext, TriggerRenderer, TriggerRendererContext } from "../types";
import type { Predicate } from "../utils";
import { formatPredicate, stringOrDash } from "../utils";
import githubIcon from "@/assets/icons/integrations/github.svg";
import type { TriggerProps } from "@/ui/trigger";
import type { BaseNodeMetadata } from "./types";
import { buildGithubSubtitle } from "./utils";

interface OnCheckRunConfiguration {
  statuses?: string[];
  conclusions?: string[];
  names?: Predicate[];
  branches?: Predicate[];
  pullRequestsOnly?: boolean;
}

interface CheckRunPullRequest {
  number?: number;
  html_url?: string;
  head?: {
    ref?: string;
    sha?: string;
  };
  base?: {
    ref?: string;
  };
}

interface CheckRun {
  name?: string;
  status?: string;
  conclusion?: string;
  head_sha?: string;
  html_url?: string;
  details_url?: string;
  app?: {
    name?: string;
  };
  check_suite?: {
    head_branch?: string;
    head_sha?: string;
  };
  pull_requests?: CheckRunPullRequest[];
}

interface OnCheckRunEventData {
  action?: string;
  check_run?: CheckRun;
  repository?: {
    full_name?: string;
    html_url?: string;
  };
  sender?: {
    login?: string;
  };
}

export const onCheckRunTriggerRenderer: TriggerRenderer = {
  getTitleAndSubtitle: (context: TriggerEventContext) => {
    const eventData = context.event?.data as OnCheckRunEventData;

    return {
      title: checkRunTitle(eventData),
      subtitle: buildGithubSubtitle(checkRunResult(eventData), context.event?.createdAt),
    };
  },

  getRootEventValues: (context: TriggerEventContext): Record<string, string> => {
    const eventData = context.event?.data as OnCheckRunEventData;

    return {
      Action: stringOrDash(eventData?.action),
      Name: stringOrDash(checkRunName(eventData)),
      Status: stringOrDash(checkRunStatus(eventData)),
      Conclusion: stringOrDash(checkRunConclusion(eventData)),
      Branch: stringOrDash(checkRunBranch(eventData)),
      SHA: stringOrDash(checkRunSha(eventData)),
      "Pull request": stringOrDash(checkRunPullRequestLabel(eventData)),
      App: stringOrDash(checkRunAppName(eventData)),
      "Details URL": stringOrDash(checkRunDetailsUrl(eventData)),
      Repository: stringOrDash(checkRunRepositoryName(eventData)),
    };
  },

  getTriggerProps: (context: TriggerRendererContext) => {
    const { node, definition, lastEvent } = context;
    const metadata = node.metadata as unknown as BaseNodeMetadata;
    const configuration = node.configuration as unknown as OnCheckRunConfiguration;

    const props: TriggerProps = {
      title: node.name || definition.label || "Unnamed trigger",
      iconSrc: githubIcon,
      iconColor: getColorClass(definition.color),
      collapsedBackground: getBackgroundColorClass(definition.color),
      metadata: checkRunMetadataItems(metadata?.repository?.name, configuration),
    };

    if (lastEvent) {
      const eventData = lastEvent.data as OnCheckRunEventData;
      props.lastEventData = {
        title: checkRunTitle(eventData),
        subtitle: buildGithubSubtitle(checkRunResult(eventData), lastEvent.createdAt),
        receivedAt: new Date(lastEvent.createdAt),
        state: "triggered",
        eventId: lastEvent.id,
      };
    }

    return props;
  },
};

function checkRunMetadataItems(
  repositoryName: string | undefined,
  configuration?: OnCheckRunConfiguration,
): MetadataItem[] {
  const metadataItems: MetadataItem[] = [];

  if (repositoryName) {
    metadataItems.push({ icon: "book", label: repositoryName });
  }

  appendCheckRunConfigurationMetadata(metadataItems, configuration);

  return metadataItems;
}

function appendCheckRunConfigurationMetadata(
  metadataItems: MetadataItem[],
  configuration?: OnCheckRunConfiguration,
): void {
  appendJoinedMetadata(metadataItems, "circle-check", configuration?.statuses);
  if (configuration?.conclusions && configuration.conclusions.length > 0) {
    metadataItems.push({ icon: "funnel", label: `conclusion ${configuration.conclusions.join(", ")}` });
  }

  if (configuration?.names && configuration.names.length > 0) {
    metadataItems.push({ icon: "funnel", label: `name ${configuration.names.map(formatPredicate).join(", ")}` });
  }

  if (configuration?.branches && configuration.branches.length > 0) {
    metadataItems.push({
      icon: "git-branch",
      label: `branch ${configuration.branches.map(formatPredicate).join(", ")}`,
    });
  }

  if (configuration?.pullRequestsOnly) {
    metadataItems.push({ icon: "git-pull-request", label: "pull requests only" });
  }
}

function appendJoinedMetadata(metadataItems: MetadataItem[], icon: string, values: string[] | undefined): void {
  if (!values || values.length === 0) {
    return;
  }

  metadataItems.push({ icon, label: values.join(", ") });
}

function checkRunTitle(eventData: OnCheckRunEventData | undefined): string {
  const name = checkRunName(eventData) || "Check run";
  const state = checkRunTitleState(checkRunStatus(eventData), checkRunConclusion(eventData));
  const sha = shortSha(checkRunSha(eventData));
  const title = state ? `${name} ${state}` : name;

  if (sha) {
    return `${title} - ${sha}`;
  }

  return title;
}

function checkRunTitleState(status: string | undefined, conclusion: string | undefined): string {
  if (status === "completed") {
    switch (conclusion) {
      case "success":
        return "succeeded";
      case "failure":
        return "failed";
      case "cancelled":
        return "was cancelled";
      case "skipped":
        return "was skipped";
      case "timed_out":
        return "timed out";
      case "action_required":
        return "needs action";
      case "stale":
        return "is stale";
      case "neutral":
        return "completed";
      default:
        return "completed";
    }
  }

  if (status === "in_progress") {
    return "is running";
  }

  if (status === "queued") {
    return "is queued";
  }

  return "";
}

function checkRunResult(eventData: OnCheckRunEventData | undefined): string {
  return checkRunConclusion(eventData) || checkRunStatus(eventData) || "";
}

function checkRunName(eventData: OnCheckRunEventData | undefined): string | undefined {
  return eventData?.check_run?.name;
}

function checkRunStatus(eventData: OnCheckRunEventData | undefined): string | undefined {
  return eventData?.check_run?.status;
}

function checkRunConclusion(eventData: OnCheckRunEventData | undefined): string | undefined {
  return eventData?.check_run?.conclusion;
}

function checkRunBranch(eventData: OnCheckRunEventData | undefined): string | undefined {
  return eventData?.check_run?.check_suite?.head_branch || eventData?.check_run?.pull_requests?.[0]?.head?.ref;
}

function checkRunSha(eventData: OnCheckRunEventData | undefined): string | undefined {
  return eventData?.check_run?.head_sha || eventData?.check_run?.check_suite?.head_sha;
}

function checkRunPullRequestLabel(eventData: OnCheckRunEventData | undefined): string | undefined {
  const pullRequest = eventData?.check_run?.pull_requests?.[0];
  if (!pullRequest?.number) {
    return undefined;
  }

  return `#${pullRequest.number}`;
}

function checkRunAppName(eventData: OnCheckRunEventData | undefined): string | undefined {
  return eventData?.check_run?.app?.name;
}

function checkRunDetailsUrl(eventData: OnCheckRunEventData | undefined): string | undefined {
  return eventData?.check_run?.details_url || eventData?.check_run?.html_url;
}

function checkRunRepositoryName(eventData: OnCheckRunEventData | undefined): string | undefined {
  return eventData?.repository?.full_name;
}

function shortSha(sha: string | undefined): string {
  return sha ? sha.slice(0, 7) : "";
}
